File: src/main/java/de/schabby/cyntaja/Struct.java
package de.schabby.cyntaja;

import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.List;

/**
 * A Struct in the C programming language.
 */
public class Struct {

    public String name;
    public final List<Field> fields = new ArrayList<>();
    private boolean isPrimitive = false;
    //private Clazz jaldaClazz = null;

    public Struct(String name)
    {
        this.name = name;
    }

    public void generateHeader(PrintWriter pw)
    {
        String hash = (""+name.hashCode()).replace("-", "A");
        pw.println("typedef struct str_"+hash+" {");
        for(Field field: fields)
        {
            pw.println("\t"+field.getVariable().fullDeclaration()+";");
        }
        pw.println("} "+name+";");
    }

    public String toC(boolean asPointerIfApplicable)
    {
        if( isPrimitive )
            return name;
        else {
            //String result = "struct " + name;
            String result = name;
            if( asPointerIfApplicable )
                result += "*";
            return result;
        }
    }

    public Field getFieldByName(String name)
    {
        for(Field f: fields)
            if( f.getVariable().identifier.equals(name) )
                return f;

        throw new RuntimeException("could not find field with name '"+name+"' in Class '"+this.name+"'");
    }

    public boolean isPrimitive()
    {
        return isPrimitive;
    }

    public void setPrimitive(boolean primitive)
    {
        isPrimitive = primitive;
    }

    /*public Clazz getJaldaClazz() {
        return jaldaClazz;
    }

    public void setJaldaClazz(Clazz jaldaClazz) {
        this.jaldaClazz = jaldaClazz;
    }*/
}

File: src/main/kotlin/de/schabby/cyntaja/Struct.kt
package de.schabby.cyntaja


import kotlin.RuntimeException

class Struct(name:String) : Writable, ValueType(name) {

    val fields = mutableListOf<Variable>()

    override fun writeCode(): String {
        val sb = StringBuilder()
        sb.append("struct $name {\n")
        fields.forEach {
            sb.append("   ")
            if( it.type is FunctionPointer ) {
                val fp = it.type as FunctionPointer
                fp.name = it.name
                sb.append(fp.writeFunctionPointer()+";\n")
            }
            else sb.append(it.type.name+" "+it.name+";\n")
        }
        sb.append("};")
        return sb.toString()
    }

    fun getFieldByVarName(varName:String) : Variable {
        val rslt = fields.find {it.name==varName}
        if( rslt == null ) throw RuntimeException("could not find field with name $varName in struct $name")
        return rslt
    }

    /**
     * Adds a new field to this Struct.
     */
    fun field(name:String, type:Type) {
        fields.add(Variable(name, type))
    }

    /**
     * Returns the field with the given name. If the
     * field is not present, this method crashes.
     */
    fun field(name:String) : Variable {
        val result = fields.filter { it.name.equals(name) }
        if(result.isEmpty()) throw RuntimeException("could not find field '$name' in Struct '${this.name}'")
        return result.first()
    }

}